qemu/target-i386/cpu.h

```diff
@@ -699,7 +699,7 @@ typedef enum {
 
 typedef struct SegmentCache {
     uint32_t selector;
-    target_ulong base;
+    uint64_t base;   /* handle 32 or 64 bit CPUs */
```

Comment:
Is it guaranteed that target_ulong is always 64 bits? If it is then why not use uint64_t? If it isn't, then we can't cast x86_mmr* to SegmentCache*.

Since the pointer that passed into the qemu code is to a struct defined in x86.h (x86_mmr) we can't cast to x86_mmr* because that type is unknown to qemu. If we cast to SegmentCache* (as is done now) then there needs to be a guarantee that the base field is the same size as the corresponding base field provided by the caller.

On 2/4/2016 5:01 PM, Nguyen Anh Quynh wrote:
> In qemu/target-i386/cpu.h <https://github.com/unicorn-engine/unicorn/pull/416#discussion_r51963462>:
> 
>> @@ -699,7 +699,7 @@ typedef enum {
>>  
>>  typedef struct SegmentCache {
>>      uint32_t selector;
>> -    target_ulong base;
>> +    uint64_t base;   /* handle 32 or 64 bit CPUs */
> 
> oh, why do we need to fix this code from Qemu? it is better to leave this unchanged, while we still expose |base| to our API as |uint64_t| for our convenience?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/unicorn-engine/unicorn/pull/416/files#r51963462>.
>